### ISSUE TITLE
Ensure each spend is broadcasted over different IP

### ISSUE THREAD

**Issue by alpha-centauri-21** [sparrowwallet/sparrow#96]
Sparrow goes to great lengths to give user as much control and understanding as possible. This is great.

One area where this is lacking is transaction broadcast. It is unclear how Sparrow broadcasts transactions.

The expected behavior:

1. Desktop wallet should force a separate Tor circuit for each spend to make sure each transaction is communicated to broadcasting node from a distinct IP (within Tor limitations / no 100% guarantee but very high likelihood).

2. Desktop wallet should use a separate broadcasting service for each spend. This is **in addition to** 1. It aims to prevent tx clustering based on the node IP. Example context: say user makes 8 spends within 10 minutes. All transactions get broadcasted by the same Bitcoin Core node (so same IP) or the same ElectrumX (again same IP). A very well connected surveillance company nodes **will** cluster these 8 spends together based on the combination of time proximity and IP address.

**Comment by craigraw** [comment]
Sparrow broadcasts transactions via the node it's connected to (optionally with an Electrum server as an intermediate).

> Desktop wallet should force a separate Tor circuit for each spend

I can't see any good reason for this? The point of using Tor is that the source IP is hidden from the destination IP.

> Desktop wallet should use a separate broadcasting service for each spend.

This would require configuring and connecting to a different node or Electrum server. One potential solution is to route through Blockstream's Electrum server .onion address, although it would require Tor to be running locally and would place a dependency on that server (I'm not aware of any other public Electrum servers available over Tor).

**Comment by alpha-centauri-21** [comment]
> Sparrow broadcasts transactions via the node it's connected to (optionally with an Electrum server as an intermediate).
> 
> > Desktop wallet should force a separate Tor circuit for each spend
> 
> I can't see any good reason for this? The point of using Tor is that the source IP is hidden from the destination IP.

It's about the IP of the Tor exit node. Yes, surveillers won't know the source IP but **they will still cluster** these spends together because they exited from the same Tor exit node within the same couple minutes.

Please note the Tor Browser does create a separate Tor circuit for each visited domain. It's the same thing.

It should be easy to do. One way to force a new circuit is it to simply use a pseudo-random `user:password` for the SOCKS5 proxy (which works regardless), like `socks5://pseudouser:pseudopass@127.0.0.1:9050` These values can be derived from spend tx, etc.

> > Desktop wallet should use a separate broadcasting service for each spend.
> 
> This would require configuring and connecting to a different node or Electrum server. One potential solution is to route through Blockstream's Electrum server .onion address, although it would require Tor to be running locally and would place a dependency on that server (I'm not aware of any other public Electrum servers available over Tor).

Transaction broadcast is conceptually decoupled from existing reliance on a single node/Electrum.

There is a number of public transaction broadcasting services that could be used. Some wallets offer "broadcast tx via X". I propose to support a couple of them and randomize by default.

**Comment by craigraw** [comment]
Implemented in 0e42c65. If there is a Tor proxy available, Sparrow will try the .onion URLs for blockstream.info and mempool.space in random order and broadcast the transaction there. Sparrow will also ensure that a new Tor circuit is created every time this is done using the proxy authentication technique you have described (I checked that this is working correctly).

Thanks for the help in understanding this better.
